test/functionality/_helper/environment-builder.ts

```diff
@@ -1,11 +1,141 @@
+	if(name === undefined) {
+		return { nodeId, name: unnamedArgument(nodeId), scope, used }
+	} else {
+		return [name, { nodeId, name, scope, used }]
+	}
+}
+
 export function unnamedArgument(id: NodeId) {
 	return `${UnnamedArgumentPrefix}${id}`
+}
+
+/**
+ * The constant global environment with all pre-defined functions.
+ */
+export const defaultEnvironment = () => {
+	const builtIns = new Map<Identifier, IdentifierDefinition[]>(DefaultEnvironmentMemory)
+	const globalEnv = new Environment(GlobalScope)
+	globalEnv.memory = builtIns
+	return new EnvironmentBuilder(globalEnv, 0)
+}
+
+/**
+ * EnvironmentBuilder extends REnvironmentInformation with builder pattern methods.
+ */
+export class EnvironmentBuilder implements REnvironmentInformation {
+	/**
+	 * Use global environment.
+	 */
+	current: Environment
+	/**
+	 * Level is 0.
+	 */
+	level:   number
+
+	constructor(env: Environment, level: number) {
+		this.current = env
+		this.level = level
+	}
+
+	/**
+	 * Defines a new argument in the top environment.
+	 * @param name - Argument name
+	 * @param definedAt - AST Node Id of definition
+	 * @param nodeId - AST Node ID of usage
+	 * @param scope - local (default) or optional
+	 * @param used - always (default) or optional
+	 */
+	defineArgument(name: string, definedAt: NodeId, nodeId: NodeId, scope: RScope = LocalScope, used: WhenUsed = 'always') {
+		return this.defineEnv({ name, kind: 'argument', definedAt, nodeId, scope, used })
+	}
+
+	/**
+	 * Defines a new function in the top environment.
+	 * @param name - Function name
+	 * @param definedAt - AST Node Id of definition
+	 * @param nodeId - AST Node ID of usage
+	 * @param scope - local (default) or optional
+	 * @param used - always (default) or optional
+	 */
+	defineFunction(name: string, definedAt: NodeId, nodeId: NodeId, scope: RScope = LocalScope, used: WhenUsed = 'always') {
+		return this.defineEnv({ name, kind: 'function', definedAt, nodeId, scope, used })
+	}
+
+	/**
+	 * Defines a new parameter in the top environment.
+	 * @param name - Parameter name
+	 * @param definedAt - AST Node Id of definition
+	 * @param nodeId - AST Node ID of usage
+	 * @param scope - local (default) or optional
+	 * @param used - always (default) or optional
+	 */
+	defineParameter(name: string, definedAt: NodeId, nodeId: NodeId, scope: RScope = LocalScope, used: WhenUsed = 'always') {
+		return this.defineEnv({ name, kind: 'parameter', definedAt, nodeId, scope, used })
+	}
+
+	/**
+	 * Defines a new parameter in the top environment.
+	 * @param name - Variable name
+	 * @param definedAt - AST Node Id of definition
+	 * @param nodeId - AST Node ID of usage; optional
+	 * @param scope - local (default) or optional
+	 * @param used - always (default) or optional
+	 */
+	defineVariable(name: string, definedAt: NodeId, nodeId: NodeId = '_0', scope: RScope = LocalScope, used: WhenUsed = 'always') {
```

Comment:
so just one last thing: i do not get the swap of `definedAt` and `nodeId` - why is (wit halle functions) the `definedAt` before the node id? The nodeId is always linked to the name.